### PR TITLE
Add reddit dataset.

### PR DESCRIPTION
**Pull request by coledie**
- [x] Forked repo is synced with upstream -> github shows no code delta outside of the desired.
    https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork
- [x] Tests are passing? https://github.com/microsoft/DeepGNN/blob/main/CONTRIBUTING.md#run-tests
- [x] Changelog and documentation updated.
- [x] PR is labeled using the label menu on the right side.

New Behavior
----------------
Add reddit dataset using same code as PPI.

### CODE REVIEW

File: src/python/deepgnn/graph_engine/data/reddit.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+def onehot(value, size):
+    """Convert value to onehot vector."""
+    output = [0] * size
+    output[value] = 1
+    return output
+
+
+class Reddit(Client):
+    """
+    Reddit Subreddit-Subreddit Interactions graph.
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/data/reddit.py:29]: Update changelog?

File: src/python/deepgnn/graph_engine/data/reddit.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+
+def onehot(value, size):
+    """Convert value to onehot vector."""
+    output = [0] * size
+    output[value] = 1
+    return output
+
+
+class Reddit(Client):
```
[the remark this answers is not in the document]
Comment (coledie) [reply src/python/deepgnn/graph_engine/data/reddit.py]: Reddit now uses all PPI functions so there is no test.

File: src/python/deepgnn/graph_engine/data/reddit.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    def __init__(self, output_dir: str = None, edge_downsample_pct: float = 0.1):
+        """
+        Initialize Reddit dataset.
+
+        Args:
+          output_dir (string): file directory for graph data.
+          edge_downsample_pct (float, default=.1): Percent of edges to use, default reddit graph has 100M edges.
+        """
+        self._edge_downsample_pct = edge_downsample_pct
+        self.url = "https://snap.stanford.edu/graphsage/reddit.zip"
```
[the remark this answers is not in the document]
Comment (coledie) [reply src/python/deepgnn/graph_engine/data/reddit.py:60]: This will not be on CI, and snap.graphsage is supposed to be a stable source.

File: src/python/deepgnn/graph_engine/data/ppi.py [hunk 4]
```diff
@@ -108,16 +118,16 @@ def _load_raw_graph(self, data_dir: str):
 
         return nodes, nodes_type, train_neighbors, other_neighbors, feats, class_map
 
-    def _build_graph(self, output_dir: str) -> str:
+    def _build_graph(self, output_dir: str, load_graph_output: tuple) -> str:
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/data/ppi.py:121]: why do we need to pass load_graph_output tuple instead of overriding it?

File: src/python/deepgnn/graph_engine/data/reddit.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+from typing import Dict
+
+import numpy as np
+from sklearn.preprocessing import StandardScaler
+
+from deepgnn.graph_engine.data.data_util import download_file
+from deepgnn.graph_engine.data.ppi import PPI
+
+
+def onehot(value, size):
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/data/reddit.py:18]: Why this method is public, do you intend to use it somewhere else?

File: src/python/deepgnn/graph_engine/data/ppi.py [hunk 6]
```diff
@@ -36,26 +36,36 @@ class PPI(Client):
     - Node Feature Dim: 50 (id:1)
     """
 
-    def __init__(self, output_dir: str = None):
+    def __init__(self, output_dir: str = None, n_partitions: int = 1):
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/data/ppi.py:39]: Please, use `num_partitions` for consistency with dispatcher arguments.

File: src/python/deepgnn/graph_engine/data/ppi.py (hunk 4 above)
Comment (coledie) [reply src/python/deepgnn/graph_engine/data/ppi.py:121]: If I call PPI's build_graph function from reddit, it will call PPI's load_graph function even if Reddit overloads load_graph. This is why I needed to refactor into constructor.

File: src/python/deepgnn/graph_engine/data/reddit.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+                train_nodes.append(nid)
+            nodes.append(nid)
+            nodes_type.append(ntype)
+
+        train_neighbors: Dict = {nid: [] for nid in id_map.values()}
+        other_neighbors: Dict = {nid: [] for nid in id_map.values()}
+
+        # edges
+        np.random.seed(0)
+        edge_downsample_mask = (
```
[the remark this answers is not in the document]
Comment (coledie) [reply src/python/deepgnn/graph_engine/data/reddit.py:104]: It takes 5-10 minutes to convert reddit with the downsampled edges already, this is good enough for benchmarking

File: CHANGELOG.md [hunk 8]
```diff
@@ -6,6 +6,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added
```
Comment (mortid0) [inline CHANGELOG.md:9]: nit: looks like this PR needs a rebase.

File: src/python/deepgnn/graph_engine/data/ppi.py (hunk 4 above)
Comment (mortid0) [reply src/python/deepgnn/graph_engine/data/ppi.py:121]: There is probably something else going on, because this code works perfectly fine:
image

File: src/python/deepgnn/graph_engine/data/reddit.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+                ntype = NODE_TYPE_ID["train"]
+                train_nodes.append(nid)
+            nodes.append(nid)
+            nodes_type.append(ntype)
+
+        train_neighbors: Dict = {nid: [] for nid in id_map.values()}
+        other_neighbors: Dict = {nid: [] for nid in id_map.values()}
+
+        # edges
+        np.random.seed(0)
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/data/reddit.py:103]: nit: you probably want to put this seed to main function and let callers to control seeds.

File: src/python/deepgnn/graph_engine/data/reddit.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+          seed (int, default=0): Seed for random number generation.
+        """
+        self._edge_downsample_pct = edge_downsample_pct
+        self._num_partitions = num_partitions
+        np.random.seed(seed)
+        self.url = "https://snap.stanford.edu/graphsage/reddit.zip"
+        self.GRAPH_NAME = "reddit"
+        self.output_dir = output_dir
+        if self.output_dir is None:
+            self.output_dir = os.path.join("/tmp/", self.GRAPH_NAME)
```
Comment (mortid0) [inline src/python/deepgnn/graph_engine/data/reddit.py:65]: `s/"/tmp/"/tempfile.gettempdir()/`

**Review by mortid0 — APPROVED** [review]

File: src/python/deepgnn/graph_engine/data/reddit.py (hunk 10 above)
Comment (coledie) [reply src/python/deepgnn/graph_engine/data/reddit.py:65]: This is a copy paste of PPI.
